File: examples/flutter/scripts/agent_mcp_http_smoke.dart
import 'dart:io';

import 'package:http/http.dart' as http;

import 'test_support.dart';

const String mcpProtocolVersion = '2025-11-25';
const String mcpLegacyProtocolVersion = '2025-06-18';

Future<({http.Response response, JsonMap body})> mcpPost(
  String endpoint,
  JsonMap payload,
) async {
  final response = await postJson(
    endpoint,
    payload,
    headers: <String, String>{'accept': 'application/json, text/event-stream'},
  );
  final body = await parseJsonResponse(response, 'MCP HTTP response');
  return (response: response, body: body);
}

Future<void> run() async {
  final tempDirectory = await createTempDataDirectory(
    'app-flutter-agent-mcp-http-smoke-',
  );
  HttpServer? backendServer;
  HttpServer? agentServer;

  try {
    final backend = await startBackendHost(tempDirectory);
    final agent = await startAgentHost(tempDirectory);
    backendServer = backend.server;
    agentServer = agent.server;

    final mcpUrl = '${agent.baseUrl}/mcp';

    final getResponse = await getJson(
      mcpUrl,
      headers: <String, String>{'accept': 'text/event-stream'},
    );
    expectEqual(
      getResponse.statusCode,
      405,
      'agent_mcp_http_smoke: GET /mcp should explicitly reject SSE when unsupported',
    );

    final invalidInitialize = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 1,
      'method': 'initialize',
      'params': <String, dynamic>{
        'protocolVersion': '1999-01-01',
        'capabilities': <String, dynamic>{'tools': <String, dynamic>{}},
        'clientInfo': <String, dynamic>{
          'name': 'agent-mcp-http-smoke',
          'version': '1.0.0',
        },
      },
    });
    expectEqual(
      invalidInitialize.response.statusCode,
      200,
      'agent_mcp_http_smoke: initialize errors should return JSON-RPC payloads',
    );
    expectTrue(
      asJsonMap(
        invalidInitialize.body['error'],
        'invalid initialize error',
      )['message'].toString().contains('Unsupported MCP protocol version'),
      'agent_mcp_http_smoke: initialize must reject unsupported MCP versions',
    );

    final initialize = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 2,
      'method': 'initialize',
      'params': <String, dynamic>{
        'protocolVersion': mcpLegacyProtocolVersion,
        'capabilities': <String, dynamic>{'tools': <String, dynamic>{}},
        'clientInfo': <String, dynamic>{
          'name': 'agent-mcp-http-smoke',
          'version': '1.0.0',
        },
      },
    });
    final initializeResult = asJsonMap(
      initialize.body['result'],
      'initialize result',
    );
    expectEqual(
      initializeResult['protocolVersion'],
      mcpProtocolVersion,
      'agent_mcp_http_smoke: initialize must negotiate the host-supported protocol version',
    );
    expectEqual(
      asJsonMap(initializeResult['serverInfo'], 'serverInfo')['name'],
      'flutter-task-board-agent',
      'agent_mcp_http_smoke: initialize must expose the agent host server name',
    );
    expectEqual(
      asJsonMap(
        asJsonMap(
          initializeResult['capabilities'],
          'capabilities',
        )['resources'],
        'resource capabilities',
      )['listChanged'],
      false,
      'agent_mcp_http_smoke: initialize must advertise MCP resources support',
    );
    expectTrue(
      initializeResult['instructions'].toString().contains('Tool task_list'),
      'agent_mcp_http_smoke: initialize instructions must include registry-derived prompt fragments',
    );

    final initializedNotification = await postJson(
      mcpUrl,
      <String, dynamic>{
        'jsonrpc': '2.0',
        'method': 'notifications/initialized',
        'params': const <String, dynamic>{},
      },
      headers: <String, String>{
        'accept': 'application/json, text/event-stream',
      },
    );
    expectEqual(
      initializedNotification.statusCode,
      202,
      'agent_mcp_http_smoke: notifications/initialized should be accepted',
    );

    final listedTools = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 3,
      'method': 'tools/list',
      'params': const <String, dynamic>{},
    });
    final listedToolsResult = asJsonMap(
      listedTools.body['result'],
      'tools/list result',
    );
    final listedToolDescriptors = asJsonMapList(
      listedToolsResult['tools'],
      'tools/list tools',
    );
    final listedToolNames =
        listedToolDescriptors
            .map((JsonMap tool) => tool['name'].toString())
            .toList(growable: false)
          ..sort();
    expectJsonEqual(
      listedToolNames,
      <String>['task_create', 'task_list', 'task_move'],
      'agent_mcp_http_smoke: remote MCP tools/list must expose the three task tools',
    );
    expectTrue(
      listedToolDescriptors.any(
        (JsonMap tool) =>
            tool['name'] == 'task_list' &&
            tool['description'].toString().contains(
              'Use when: When the user asks to see the board',
            ),
      ),
      'agent_mcp_http_smoke: remote MCP descriptors must include semantic summaries from prompt/discovery/context',
    );

    final listedResources = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 4,
      'method': 'resources/list',
      'params': const <String, dynamic>{},
    });
    final listedResourcesResult = asJsonMap(
      listedResources.body['result'],
      'resources/list result',
    );
    final resourceUris =
        asJsonMapList(
              listedResourcesResult['resources'],
              'resources/list resources',
            )
            .map((JsonMap resource) => resource['uri'].toString())
            .toList(growable: false)
          ..sort();
    expectJsonEqual(
      resourceUris,
      <String>[
        'app://agent/system/prompt',
        'app://agent/tools/task_create/semantic',
        'app://agent/tools/task_list/semantic',
        'app://agent/tools/task_move/semantic',
      ],
      'agent_mcp_http_smoke: remote MCP resources/list must publish the system prompt plus per-tool semantic resources',
    );

    final promptResource = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 5,
      'method': 'resources/read',
      'params': <String, dynamic>{'uri': 'app://agent/system/prompt'},
    });
    final promptResult = asJsonMap(
      promptResource.body['result'],
      'resources/read result',
    );
    final promptText = asJsonMapList(
      promptResult['contents'],
      'prompt contents',
    ).first['text'].toString();
    expectTrue(
      promptText.contains('Tool task_move'),
      'agent_mcp_http_smoke: resources/read must expose the host global prompt built from tool fragments',
    );

    final invalidCreate = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 6,
      'method': 'tools/call',
      'params': <String, dynamic>{
        'name': 'task_create',
        'arguments': <String, dynamic>{'title': '   '},
      },
    });
    final invalidCreateResult = asJsonMap(
      invalidCreate.body['result'],
      'invalid create result',
    );
    expectEqual(
      invalidCreateResult['isError'],
      true,
      'agent_mcp_http_smoke: invalid task_create must fail through remote MCP',
    );
    expectEqual(
      asJsonMap(
        asJsonMap(
          invalidCreateResult['structuredContent'],
          'invalid create structuredContent',
        )['error'],
        'invalid create error',
      )['code'],
      'VALIDATION_FAILED',
      'agent_mcp_http_smoke: invalid task_create must preserve the APP validation code',
    );

    final created = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 7,
      'method': 'tools/call',
      'params': <String, dynamic>{
        'name': 'task_create',
        'arguments': <String, dynamic>{
          'title': 'Remote MCP task',
          'description': 'Created through streamable HTTP MCP',
        },
      },
    });
    final createdResult = asJsonMap(created.body['result'], 'created result');
    expectEqual(
      createdResult['isError'],
      false,
      'agent_mcp_http_smoke: task_create must succeed through remote MCP',
    );
    final createdTask = asJsonMap(
      asJsonMap(
        createdResult['structuredContent'],
        'created structuredContent',
      )['task'],
      'created task',
    );

    final moveWithoutConfirmation = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 8,
      'method': 'tools/call',
      'params': <String, dynamic>{
        'name': 'task_move',
        'arguments': <String, dynamic>{
          'taskId': createdTask['id'],
          'targetStatus': 'doing',
        },
      },
    });
    final moveWithoutConfirmationResult = asJsonMap(
      moveWithoutConfirmation.body['result'],
      'move without confirmation result',
    );
    expectEqual(
      moveWithoutConfirmationResult['isError'],
      true,
      'agent_mcp_http_smoke: task_move without confirmation must fail through remote MCP',
    );
    expectEqual(
      asJsonMap(
        asJsonMap(
          moveWithoutConfirmationResult['structuredContent'],
          'move without confirmation structuredContent',
        )['error'],
        'move without confirmation error',
      )['code'],
      'CONFIRMATION_REQUIRED',
      'agent_mcp_http_smoke: task_move without confirmation must preserve the confirmation code',
    );

    final moved = await mcpPost(mcpUrl, <String, dynamic>{
      'jsonrpc': '2.0',
      'id': 9,
      'method': 'tools/call',
      'params': <String, dynamic>{
        'name': 'task_move',
        'arguments': <String, dynamic>{
          'taskId': createdTask['id'],
          'targetStatus': 'doing',
          'confirmed': true,
        },
      },
    });
    final movedResult = asJsonMap(moved.body['result'], 'moved result');
    expectEqual(
      movedResult['isError'],
      false,
      'agent_mcp_http_smoke: confirmed task_move must succeed through remote MCP',
    );

    final backendTasks = await assertSuccessEnvelope(
      await getJson('${backend.baseUrl}/tasks'),
      'backend list after remote MCP execution',
    );
    final backendTaskList = asJsonMapList(
      backendTasks['tasks'],
      'backend tasks after remote MCP execution',
    );
    final backendCreatedTask = backendTaskList.firstWhere(
      (JsonMap task) => task['id'] == createdTask['id'],
      orElse: () => fail(
        'agent_mcp_http_smoke: backend must observe tasks created through remote MCP',
      ),
    );
    expectEqual(
      backendCreatedTask['status'],
      'doing',
      'agent_mcp_http_smoke: backend must observe status changes made through remote MCP',
    );

    stdout.writeln('agent_mcp_http_smoke: ok');
  } finally {
    await closeServer(agentServer);
    await closeServer(backendServer);
    await removeTempDirectory(tempDirectory);
  }
}

Future<void> main() async {
  try {
    await run();
  } catch (error) {
    stderr.writeln(error);
    exitCode = 1;
  }
}

File: examples/next/scripts/agent_mcp_http_smoke.ts
import assert from "node:assert/strict";
import { mkdtemp, rm } from "node:fs/promises";
import type { Server } from "node:http";
import os from "node:os";
import { join } from "node:path";

import { bootstrap as bootstrapAgent } from "../apps/agent/app";
import { bootstrap as bootstrapBackend } from "../apps/backend/app";

const MCP_PROTOCOL_VERSION = "2025-11-25";
const MCP_LEGACY_PROTOCOL_VERSION = "2025-06-18";

type JsonRpcResponse = {
  jsonrpc: "2.0";
  id: string | number | null;
  result?: unknown;
  error?: {
    code: number;
    message: string;
    data?: unknown;
  };
};

type Task = {
  id: string;
  title: string;
  description?: string;
  status: "todo" | "doing" | "done";
  createdAt: string;
  updatedAt: string;
};

type McpToolCallResult = {
  content: Array<{ type: "text"; text: string }>;
  structuredContent?: unknown;
  isError?: boolean;
};

async function closeServer(server: Server): Promise<void> {
  await new Promise<void>((resolve, reject) => {
    server.close((error) => {
      if (error) {
        reject(error);
        return;
      }

      resolve();
    });
  });
}

async function startServer(
  kind: "backend" | "agent",
  dataDirectory: string
): Promise<{
  server: Server;
  baseUrl: string;
}> {
  if (kind === "backend") {
    const app = bootstrapBackend({
      dataDirectory,
      port: 0,
    });
    const server = await app.startBackend();
    const address = server.address();

    if (!address || typeof address === "string") {
      throw new Error("agent_mcp_http_smoke: failed to resolve backend address");
    }

    return {
      server,
      baseUrl: `http://127.0.0.1:${address.port}`,
    };
  }

  const app = bootstrapAgent({
    dataDirectory,
    port: 0,
  });
  const server = await app.startAgent();
  const address = server.address();

  if (!address || typeof address === "string") {
    throw new Error("agent_mcp_http_smoke: failed to resolve agent address");
  }

  return {
    server,
    baseUrl: `http://127.0.0.1:${address.port}`,
  };
}

async function mcpPost<T>(
  endpoint: string,
  payload: unknown
): Promise<{
  response: Response;
  body: T;
}> {
  const response = await fetch(endpoint, {
    method: "POST",
    headers: {
      accept: "application/json, text/event-stream",
      "content-type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  const body = (await response.json()) as T;
  return { response, body };
}

async function assertBackendTasks(
  baseUrl: string
): Promise<{ tasks: Task[] }> {
  const response = await fetch(`${baseUrl}/tasks`);
  const payload = (await response.json()) as {
    success: boolean;
    data?: { tasks: Task[] };
  };

  if (!response.ok || !payload.success || !payload.data) {
    throw new Error("agent_mcp_http_smoke: backend list failed");
  }

  return payload.data;
}

async function run(): Promise<void> {
  const tempDirectory = await mkdtemp(
    join(os.tmpdir(), "app-next-agent-mcp-http-smoke-")
  );
  let backendServer: Server | undefined;
  let agentServer: Server | undefined;

  try {
    const backend = await startServer("backend", tempDirectory);
    const agent = await startServer("agent", tempDirectory);
    backendServer = backend.server;
    agentServer = agent.server;

    const mcpUrl = `${agent.baseUrl}/mcp`;

    const getResponse = await fetch(mcpUrl, {
      headers: {
        accept: "text/event-stream",
      },
    });
    assert.equal(
      getResponse.status,
      405,
      "agent_mcp_http_smoke: GET /mcp should explicitly reject SSE when unsupported"
    );

    const invalidInitialize = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 1,
      method: "initialize",
      params: {
        protocolVersion: "1999-01-01",
        capabilities: {
          tools: {},
        },
        clientInfo: {
          name: "agent-mcp-http-smoke",
          version: "1.0.0",
        },
      },
    });

    assert.equal(
      invalidInitialize.response.status,
      200,
      "agent_mcp_http_smoke: initialize errors should return JSON-RPC payloads"
    );
    assert.equal(
      invalidInitialize.body.error?.message.includes(
        "Unsupported MCP protocol version"
      ),
      true,
      "agent_mcp_http_smoke: initialize must reject unsupported MCP versions"
    );

    const initialize = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 2,
      method: "initialize",
      params: {
        protocolVersion: MCP_LEGACY_PROTOCOL_VERSION,
        capabilities: {
          tools: {},
        },
        clientInfo: {
          name: "agent-mcp-http-smoke",
          version: "1.0.0",
        },
      },
    });

    assert.equal(
      initialize.body.result !== undefined,
      true,
      "agent_mcp_http_smoke: initialize must return a result"
    );
    const initializeResult = initialize.body.result as {
      protocolVersion: string;
      capabilities?: {
        resources?: {
          listChanged?: boolean;
        };
      };
      serverInfo: { name: string };
      instructions?: string;
    };
    assert.equal(
      initializeResult.protocolVersion,
      MCP_PROTOCOL_VERSION,
      "agent_mcp_http_smoke: initialize must negotiate the host-supported protocol version"
    );
    assert.equal(
      initializeResult.serverInfo.name,
      "next-task-board-agent",
      "agent_mcp_http_smoke: initialize must expose the agent host server name"
    );
    assert.equal(
      initializeResult.capabilities?.resources?.listChanged,
      false,
      "agent_mcp_http_smoke: initialize must advertise MCP resources support"
    );
    assert.equal(
      initializeResult.instructions?.includes("Tool task_list"),
      true,
      "agent_mcp_http_smoke: initialize instructions must include registry-derived prompt fragments"
    );

    const initializedNotification = await fetch(mcpUrl, {
      method: "POST",
      headers: {
        accept: "application/json, text/event-stream",
        "content-type": "application/json",
      },
      body: JSON.stringify({
        jsonrpc: "2.0",
        method: "notifications/initialized",
        params: {},
      }),
    });

    assert.equal(
      initializedNotification.status,
      202,
      "agent_mcp_http_smoke: notifications/initialized should be accepted"
    );

    const listed = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 3,
      method: "tools/list",
      params: {},
    });

    const listedTools = (
      listed.body.result as {
        tools: Array<{ name: string; description?: string }>;
      }
    ).tools
      .map((tool) => tool.name)
      .sort();
    assert.deepEqual(
      listedTools,
      ["task_create", "task_list", "task_move"],
      "agent_mcp_http_smoke: remote MCP tools/list must expose the three task tools"
    );
    assert.equal(
      (
        listed.body.result as {
          tools: Array<{ name: string; description?: string }>;
        }
      ).tools.some(
        (tool) =>
          tool.name === "task_list" &&
          tool.description?.includes("Use when: When the user asks to see the board")
      ),
      true,
      "agent_mcp_http_smoke: remote MCP descriptors must include semantic summaries from prompt/discovery/context"
    );

    const listedResources = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 4,
      method: "resources/list",
      params: {},
    });
    const resourceUris = (
      listedResources.body.result as {
        resources: Array<{ uri: string }>;
      }
    ).resources
      .map((resource) => resource.uri)
      .sort();
    assert.deepEqual(
      resourceUris,
      [
        "app://agent/system/prompt",
        "app://agent/tools/task_create/semantic",
        "app://agent/tools/task_list/semantic",
        "app://agent/tools/task_move/semantic",
      ],
      "agent_mcp_http_smoke: remote MCP resources/list must publish the system prompt plus per-tool semantic resources"
    );

    const promptResource = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 5,
      method: "resources/read",
      params: {
        uri: "app://agent/system/prompt",
      },
    });
    const promptText = (
      promptResource.body.result as { contents: Array<{ text: string }> }
    ).contents[0]?.text;
    assert.equal(
      promptText?.includes("Tool task_move"),
      true,
      "agent_mcp_http_smoke: resources/read must expose the host global prompt built from tool fragments"
    );

    const invalidCreate = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 6,
      method: "tools/call",
      params: {
        name: "task_create",
        arguments: {
          title: "   ",
        },
      },
    });

    const invalidCreateResult = invalidCreate.body.result as McpToolCallResult;
    assert.equal(
      invalidCreateResult.isError,
      true,
      "agent_mcp_http_smoke: invalid task_create must fail through remote MCP"
    );
    assert.equal(
      (invalidCreateResult.structuredContent as {
        error?: { code?: string };
      }).error?.code,
      "VALIDATION_FAILED",
      "agent_mcp_http_smoke: invalid task_create must preserve the APP validation code"
    );

    const created = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 7,
      method: "tools/call",
      params: {
        name: "task_create",
        arguments: {
          title: "Remote MCP task",
          description: "Created through streamable HTTP MCP",
        },
      },
    });

    const createdResult = created.body.result as McpToolCallResult;
    assert.equal(
      createdResult.isError,
      false,
      "agent_mcp_http_smoke: task_create must succeed through remote MCP"
    );
    const createdTask = (createdResult.structuredContent as { task: Task }).task;

    const moveWithoutConfirmation = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 8,
      method: "tools/call",
      params: {
        name: "task_move",
        arguments: {
          taskId: createdTask.id,
          targetStatus: "doing",
        },
      },
    });

    const moveWithoutConfirmationResult =
      moveWithoutConfirmation.body.result as McpToolCallResult;
    assert.equal(
      moveWithoutConfirmationResult.isError,
      true,
      "agent_mcp_http_smoke: task_move without confirmation must fail through remote MCP"
    );
    assert.equal(
      (moveWithoutConfirmationResult.structuredContent as {
        error?: { code?: string };
      }).error?.code,
      "CONFIRMATION_REQUIRED",
      "agent_mcp_http_smoke: task_move without confirmation must preserve the confirmation code"
    );

    const moved = await mcpPost<JsonRpcResponse>(mcpUrl, {
      jsonrpc: "2.0",
      id: 9,
      method: "tools/call",
      params: {
        name: "task_move",
        arguments: {
          taskId: createdTask.id,
          targetStatus: "doing",
          confirmed: true,
        },
      },
    });

    const movedResult = moved.body.result as McpToolCallResult;
    assert.equal(
      movedResult.isError,
      false,
      "agent_mcp_http_smoke: confirmed task_move must succeed through remote MCP"
    );

    const backendTasks = await assertBackendTasks(backend.baseUrl);
    assert.equal(
      backendTasks.tasks.some((task) => task.id === createdTask.id),
      true,
      "agent_mcp_http_smoke: backend must observe tasks created through remote MCP"
    );
    assert.equal(
      backendTasks.tasks.find((task) => task.id === createdTask.id)?.status,
      "doing",
      "agent_mcp_http_smoke: backend must observe status changes made through remote MCP"
    );
  } finally {
    await Promise.allSettled([
      backendServer ? closeServer(backendServer) : Promise.resolve(),
      agentServer ? closeServer(agentServer) : Promise.resolve(),
    ]);
    await rm(tempDirectory, {
      recursive: true,
      force: true,
    });
  }
}

void run().catch((error: unknown) => {
  console.error(
    "agent_mcp_http_smoke failed",
    error instanceof Error ? error.message : error
  );
  process.exitCode = 1;
});
